test/test-fork.c

```diff
@@ -0,0 +1,574 @@
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+/* these tests are Unix only */
+#ifndef _WIN32
+
+#include <unistd.h>
+#include <sys/wait.h>
+#include <sys/socket.h>
+#include <string.h>
+
+#include "uv.h"
+#include "task.h"
+
+static int timer_cb_called;
+static int socket_cb_called;
+
+static void timer_cb(uv_timer_t* timer) {
+  timer_cb_called++;
+  uv_close((uv_handle_t*) timer, NULL);
+}
+
+static int socket_cb_read_fd;
+static int socket_cb_read_size;
+static char socket_cb_read_buf[1024];
+
+static void socket_cb(uv_poll_t* poll, int status, int events) {
+  ssize_t cnt = 0;
+  socket_cb_called++;
+  ASSERT(0 == status);
+  printf("Socket cb got events %d\n", events);
+  ASSERT(UV_READABLE == (events & UV_READABLE));
+  if (socket_cb_read_fd) {
+    cnt = read(socket_cb_read_fd, socket_cb_read_buf, socket_cb_read_size);
+    ASSERT(cnt == socket_cb_read_size);
+  }
+  uv_close((uv_handle_t*) poll, NULL);
+}
+
+static void run_timer_loop_once(void) {
+  uv_loop_t* loop;
+  uv_timer_t timer_handle;
+
+  loop = uv_default_loop();
+
+  timer_cb_called = 0; /* reset for the child */
+
+  ASSERT(0 == uv_timer_init(loop, &timer_handle));
+  ASSERT(0 == uv_timer_start(&timer_handle, timer_cb, 1, 0));
+  ASSERT(0 == uv_run(loop, UV_RUN_DEFAULT));
+  ASSERT(1 == timer_cb_called);
+}
+
+static void assert_wait_child(pid_t child_pid) {
+  pid_t waited_pid;
+  int child_stat;
+
+  waited_pid = waitpid(child_pid, &child_stat, 0);
+  printf("Waited pid is %d with status %d\n", waited_pid, child_stat);
+  if (waited_pid == -1) {
+    perror("Failed to wait");
+  }
+  ASSERT(child_pid == waited_pid);
+  ASSERT(WIFEXITED(child_stat)); /* clean exit, not a signal */
+  ASSERT(!WIFSIGNALED(child_stat));
+  ASSERT(0 == WEXITSTATUS(child_stat));
+}
+
+TEST_IMPL(fork_timer) {
+  /* timers continue to work after we fork */
+
+  /*
+   * Establish the loop before we fork to make sure that it
+   * has state to get reset after the fork.
+   */
+  pid_t child_pid;
+
+  run_timer_loop_once();
+  child_pid = fork();
+  ASSERT(child_pid != -1);
+
+  if (child_pid != 0) {
+    /* parent */
+    assert_wait_child(child_pid);
+  } else {
+    /* child */
+    ASSERT(0 == uv_loop_fork(uv_default_loop()));
+    run_timer_loop_once();
+    exit(0);
```

Comment:
We have asserts that the child exited with status 0. I prefer to make that exit status explicit rather than relying on the magic of falling out of the C runtime to set that. 

It's also necessary so that any code after the brackets of the else is not run in the child but only the parent.